plugin-energy-attribution/src/transform.rs

```diff
@@ -0,0 +1,154 @@
+    fn buffer_bouncer(&mut self, measurements: &mut alumet::measurement::MeasurementBuffer) {
+        // Retrieving the metric_id of the energy attribution.
+        // Using a nested scope to reduce the lock time.
+        let metric_id = {
+            let metrics = self.metrics.lock().unwrap();
+            metrics.pod_attributed_energy.unwrap()
+        };
+
+        // If the buffers do have enough (every) MeasurementPoints,
+        // then we compute the energy attribution.
+        while self.buffer_rapl.len() >= 2 && self.buffer_pod.len() >= 2 {
+            // Get the smallest rapl id i.e. the oldest timestamp (key) present in the buffer.
+            let rapl_mini_id = self
+                .buffer_rapl
+                .keys()
+                .reduce(|x, y| if x < y { x } else { y })
+                .unwrap()
+                .clone();
+
+            // Check if the buffer_pod contains the key to prevent any panic/error bellow.
+            if !self.buffer_pod.contains_key(&rapl_mini_id) {
+                todo!("decide what to do in this case");
+            }
+
+            let rapl_point = self.buffer_rapl.remove(&rapl_mini_id).unwrap();
+
+            // Compute the sum of every `total_usage_usec` for the given timestamp: `rapl_mini_id`.
+            let tot_time_sum = self
+                .buffer_pod
+                .get(&rapl_mini_id)
+                .unwrap()
+                .iter()
+                .map(|x| match x.value {
+                    WrappedMeasurementValue::F64(fx) => fx,
+                    WrappedMeasurementValue::U64(ux) => ux as f64,
+                })
+                .sum::<f64>();
+
+            // Then for every points in the buffer_pod at `rapl_mini_id`.
+            for point in self.buffer_pod.remove(&rapl_mini_id).unwrap().iter() {
+                // We extract the current tot_time as f64.
+                let cur_tot_time_f64 = match point.value {
+                    WrappedMeasurementValue::F64(fx) => fx,
+                    WrappedMeasurementValue::U64(ux) => ux as f64,
+                };
+
+                // We create the new MeasurementPoint for the energy attribution.
+                let mut new_m = MeasurementPoint::new(
+                    rapl_point.timestamp,
+                    metric_id,
+                    point.resource.clone(),
+                    point.consumer.clone(),
+                    match rapl_point.value {
+                        WrappedMeasurementValue::F64(fx) => cur_tot_time_f64 / tot_time_sum * fx,
+                        WrappedMeasurementValue::U64(ux) => cur_tot_time_f64 / tot_time_sum * (ux as f64),
+                    },
+                );
+
+                // Then we copy every attributes of the pod to the new MeasurementPoint.
+                point
+                    .clone()
+                    .attributes()
+                    .for_each(|(key, value)| new_m = new_m.clone().with_attr(key.to_owned(), value.clone()));
```

Comment:
I won't request big changes because we already know what to improve for the next version, but I still have one comment here. Why do you clone `new_m` for each attribute? Also, you can use `with_attr_vec` or `with_attr_map` to add multiple attributes at once.

Reply:
That's a really good question, and I don't remember the answer. Because we made this with @AngeCyp few months ago 😅